### PR TITLE
refactor: simplify Uploader.vue by splitting different views into child components and purify implementations of upload

### PR DESCRIPTION
**Pull request by 599316527**
* [x] Controls as component
* [x] Separated views of Media and File Uploader
* [x] Upload implementations without UI dependencies
* [x] Add KeyField to enhance value item match
* [x] Better internal interaction between UI components and File objects
* [x] Refactor demo page of Uploader for better preview among different options
* [x] Refactor devServer for upload APIs to provide real upload experience
* [x] Update test cases (with real network requests)

### CODE REVIEW

File: packages/veui/src/mixins/upload.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  props: {
+    files: Array,
+    addable: Boolean,
+    disabled: Boolean,
+
+    // the provide and inject bindings are NOT reactive
+    options: Object
+  },
+  computed
+}
```
[the remark this answers is not in the document]
Comment (599316527) [reply packages/veui/src/mixins/upload.js:30]: 试了，inject 的 object 里面的 field 也没法 reactive

File: packages/veui/src/components/Uploader/_FileUploader.vue [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            v-if="file.status !== 'uploading'"
+            :name="icons.file"
+            :class="{
+              [$c('uploader-list-file-icon')]: true,
+              [$c('uploader-list-file-icon-failure')]:
+                file.status === 'failure'
+            }"
+          />
+          <veui-icon
+            v-if="file.status === 'uploading'"
```
[the remark this answers is not in the document]
Comment (599316527) [reply packages/veui/src/components/Uploader/_FileUploader.vue:52]: 哦哦 是的 这里是搞 UploaderFile 之前拆的，还没改到

File: packages/veui/src/components/Uploader/Uploader.vue [hunk 3]
```diff
@@ -0,0 +1,1080 @@
...
+    :is="`veui-uploader-${isMediaType ? 'media' : 'file'}`"
+    :files="fileList"
+    :addable="!canNotAddImage"
+    :disabled="realUneditable"
+    :options="childOptions"
+    @add="handleImageAdd"
+    @replace="handleImageReplace"
+    @remove="handleImageRemove"
+    @preview="handelImagePreview"
+    @custom="handleImageCustomEvent"
```
[the remark this answers is not in the document]
Comment (599316527) [reply packages/veui/src/components/Uploader/Uploader.vue:30]: 感觉还是得有，对于动作有个明确的主体
叫 Item 更合适，代表了 file, image, video, media

File: packages/veui/src/components/Uploader/Uploader.vue [hunk 4]
```diff
@@ -0,0 +1,1080 @@
...
+    ref="fileInput"
+    type="file"
+    :accept="realAccept"
+    hidden
+  >
+
+  <component
+    :is="`veui-uploader-${isMediaType ? 'media' : 'file'}`"
+    :files="fileList"
+    :addable="!canNotAddImage"
```
[the remark this answers is not in the document]
Comment (599316527) [reply packages/veui/src/components/Uploader/Uploader.vue:23]: 也是直接复用原来的computed逻辑，还没改

File: packages/veui/src/components/Uploader/Uploader.vue (hunk 3 above)
[the remark this answers is not in the document]
Comment (Justineo) [reply packages/veui/src/components/Uploader/Uploader.vue:30]: ok

File: packages/veui/src/components/Uploader/Uploader.vue [hunk 5]
```diff
@@ -0,0 +1,1080 @@
...
+        ret[key] = (...args) =>
+          compact([this[key].bind(this), callbacks[key]]).forEach(execute =>
+            execute(...args)
+          )
+        return ret
+      },
+      {}
+    )
+    this._cancel = context.uploadRequest(this._file, listeners)
+  }
```
[the remark this answers is not in the document]
Comment (599316527) [reply packages/veui/src/components/Uploader/Uploader.vue]: 正常逻辑的话应该是：

```js
    upload (options, callbacks) {
        const uploadRequest = getUploadRequest(options)
```

但这样每次 upload 都会创建一个函数出来。而这个函数其实只跟 props 里面的那些上传参数有关，写成 computed:uploadRequest 就可以利用 vue 机制来缓存上传函数，只有 props 变化时才更新

另外，把上传、取消上传的实现封装在 UploaderFile 里面，这样在 Vue 里面就只用负责视图逻辑

**Review by Justineo — CHANGES_REQUESTED** [review]

**Review by Justineo — APPROVED** [review]

**Comment by Justineo** [comment]
💯

**Pull request merged**